hoomd/hpmc/PairPotentialAngularStep.cc

```diff
@@ -0,0 +1,154 @@
+
+    if (v.is_none())
+        {
+        m_directors[particle_type_id].clear();
+        m_deltas[particle_type_id].clear();
+        return;
+        }
+    pybind11::list directors = v["directors"];
+    pybind11::list deltas = v["deltas"];
+
+    auto N = pybind11::len(directors);
+
+    if (pybind11::len(deltas) != N)
+        {
+        throw std::runtime_error("the length of the delta list should match the length"
+                                 "of the director list.");
+        }
+
+    m_directors[particle_type_id].resize(N);
+    m_deltas[particle_type_id].resize(N);
+
+    for (unsigned int i = 0; i < N; i++)
+        {
+        pybind11::tuple director_python = directors[i];
+        if (pybind11::len(director_python) != 3)
+            {
+            throw std::length_error("director must be a list of 3-tuples.");
+            }
+        m_directors[particle_type_id][i] = vec3<LongReal>(director_python[0].cast<LongReal>(),
+                                                          director_python[1].cast<LongReal>(),
+                                                          director_python[2].cast<LongReal>());
+        // TO DO: normalize m_director to be unit length
+
+        pybind11::handle delta_python = deltas[i];
+        m_deltas[particle_type_id][i] = delta_python.cast<LongReal>();
+        }
+    }
+
+pybind11::object PairPotentialAngularStep::getPatch(std::string particle_type)
+    {
+    unsigned int particle_type_id = m_sysdef->getParticleData()->getTypeByName(particle_type);
+    size_t N = m_directors[particle_type_id].size();
+
+    if (N == 0)
+        {
+        return pybind11::none();
+        }
+
+    pybind11::list directors;
+    pybind11::list deltas;
+
+    for (unsigned int i = 0; i < N; i++)
+        {
+        directors.append(pybind11::make_tuple(m_directors[particle_type_id][i].x,
+                                              m_directors[particle_type_id][i].y,
+                                              m_directors[particle_type_id][i].z));
+        deltas.append(m_deltas[particle_type_id][i]);
+        }
+
+    pybind11::dict v;
+    v["directors"] = directors;
+    v["deltas"] = deltas;
+    return v;
+    }
+
+// protected
+bool PairPotentialAngularStep::maskingFunction(LongReal r_squared,
+                                               const vec3<LongReal>& r_ij,
+                                               const unsigned int type_i,
+                                               const quat<LongReal>& q_i,
+                                               const unsigned int type_j,
+                                               const quat<LongReal>& q_j) const
+    {
+    vec3<LongReal> rhat_ij = r_ij / fast::sqrt(r_squared);
+
+    for (int m = 0; m < m_directors[type_i].size(); m++)
+        {
+        LongReal cos_delta_m = cos(m_deltas[type_i][m]);
+        vec3<LongReal> ehat_m = rotate(q_i, m_directors[type_i][m]);
```

Comment:
Pre-compute `cos_delta` in `setPatch`. The calls to `cos` in the inner loop are slowing performance.